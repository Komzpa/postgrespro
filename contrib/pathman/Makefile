# contrib/pathman/Makefile

MODULE_big = pathman
OBJS = init.o pathman.o dsm_array.o $(WIN32RES)

EXTENSION = pathman
EXTVERSION = 0.1
DATA_built = $(EXTENSION)--$(EXTVERSION).sql
PGFILEDESC = "pathman - partitioning tool"

REGRESS = pathman
EXTRA_CLEAN = $(EXTENSION)--$(EXTVERSION).sql

ifdef USE_PGXS
PG_CONFIG = pg_config
PGXS := $(shell $(PG_CONFIG) --pgxs)
include $(PGXS)
else
subdir = contrib/pathman
top_builddir = ../..
include $(top_builddir)/src/Makefile.global
include $(top_srcdir)/contrib/contrib-global.mk
endif

$(EXTENSION)--$(EXTVERSION).sql: sql/init.sql sql/hash.sql sql/range.sql
	cat $^ > $@
