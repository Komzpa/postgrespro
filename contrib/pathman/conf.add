shared_preload_libraries='pathman'
