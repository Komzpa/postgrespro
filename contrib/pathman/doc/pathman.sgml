<!-- doc/src/sgml/pathman.sgml -->

<sect1 id="pathman" xreflabel="pathman">
 <title>pathman</title>

 <indexterm zone="pathman">
  <primary>pathman</primary>
 </indexterm>

 <para>
  The <filename>pathman</filename> module provides functions ...
 </para>

 <sect2>
  <title>Pathman Concepts</title>

  <para>
   ...
  </para>
 </sect2>

 <sect2>
  <title>Functions</title>

  <para>
   The functions provided by the <filename>pathman</filename> module
   are shown in <xref linkend="pathman-func-table">.
  </para>

  <table id="pathman-func-table">
   <title><filename>pathman</filename> Functions</title>
   <tgroup cols="3">
    <thead>
     <row>
      <entry>Function</entry>
      <entry>Description</entry>
     </row>
    </thead>

    <tbody>
     <row>
      <entry><function>create_hash_partitions(text, text, integer)</function><indexterm><primary>create_hash_partitions</primary></indexterm></entry>
      <entry>
       Creates child partitions for HASH partitioning and trigger on insert. Arguments specifies parent relation name, attribute name and number of desired partitions.
      </entry>
     </row>

     <row>
      <entry><function>create_range_partitions(text, text, anyelement, anyelement, integer)</function><indexterm><primary>create_range_partitions</primary></indexterm></entry>
      <entry>
       Creates child partitions for RANGE partitioning and trigger on insert. Arguments specifies parent relation name, attribute name, start value, interval and number of premade partitions.
      </entry>
     </row>

     <row>
      <entry><function>create_range_partitions(text, text, anyelement, interval, integer)</function><indexterm><primary>create_range_partitions</primary></indexterm></entry>
      <entry>
       Same as above but suited for date or timestamp based partitioning.
      </entry>
     </row>

     <row>
      <entry><function>partition_data(text)</function><indexterm><primary>partition_data</primary></indexterm></entry>
      <entry>
       Copy data from parent table to partitions.
      </entry>
     </row>

     <row>
      <entry><function>disable_partitioning(text)</function><indexterm><primary>disable_partitioning</primary></indexterm></entry>
      <entry>
       Disables pathman for specified relation and removes the trigger.
      </entry>
     </row>

     <row>
      <entry><function>split_range_partition(text, anyelement)</function><indexterm><primary>split_range_partition</primary></indexterm></entry>
      <entry>
       Splits the RANGE partition by the specified value.
      </entry>
     </row>

     <row>
      <entry><function>split_range_partition(text, anyelement)</function><indexterm><primary>split_range_partition</primary></indexterm></entry>
      <entry>
       Merge two adjacent RANGE partitions.
      </entry>
     </row>

    </tbody>
   </tgroup>
  </table>
 </sect2>

 <sect2>
  <title>Examples</title>

  <para>
  ...
<programlisting>
CREATE TABLE words AS SELECT word FROM
        ts_stat('SELECT to_tsvector(''simple'', bodytext) FROM documents');
</programlisting>
  </para>

  <note>
   <para>
   ...
   </para>
  </note>
 </sect2>


 <sect2>
  <title>Authors</title>

  <para>
   Ildar Musin <email>i.musin@postgrespro.ru</email>, Moscow, Postgres Professional Ltd., Russia
  </para>
  <para>
   This module is sponsored by Postgres Professional Ltd., Moscow, Russia.
  </para>
 </sect2>

</sect1>
