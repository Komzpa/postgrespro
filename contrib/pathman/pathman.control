# pathman extension
comment 'Partitioning tool'
default_version = '0.1'
module_pathname='$libdir/pg_pathman'
