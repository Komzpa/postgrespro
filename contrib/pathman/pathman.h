#include "postgres.h"
#include "utils/date.h"
#include "utils/hsearch.h"
#include "nodes/pg_list.h"
#include "storage/dsm.h"
#include "storage/lwlock.h"


#define ALL NIL
#define MAX_PARTITIONS 1024

#define OP_STRATEGY_LT 1
#define OP_STRATEGY_LE 2
#define OP_STRATEGY_EQ 3
#define OP_STRATEGY_GE 4
#define OP_STRATEGY_GT 5

#define BLOCKS_COUNT 10240

/*
 * Partitioning type
 */
typedef enum PartType
{
	PT_HASH = 1,
	PT_RANGE
} PartType;

/*
 * Attribute type
 */
typedef enum AttType
{
	AT_INT = 1,
	AT_DATE,
} AttType;

typedef union rng_type
{
	int		integer;
	DateADT	date;
} rng_type;


/*
 * Dynamic shared memory array
 */
typedef struct DsmArray
{
	dsm_handle	segment;
	size_t		offset;
	size_t		length;
} DsmArray;

typedef struct Block
{
	dsm_handle	segment;
	size_t		offset;
	bool		is_free;
} Block;

typedef struct Table
{
	dsm_handle	segment;
	Block	blocks[BLOCKS_COUNT];
	size_t	block_size;
	size_t	first_free;
} Table;


/*
 * PartRelationInfo
 *		Per-relation partitioning information
 *
 *		oid - parent table oid
 *		children - list of children oids
 *		parttype - partitioning type (HASH, LIST or RANGE)
 *		attnum - attribute number of parent relation
 */
typedef struct PartRelationInfo
{
	Oid			oid;
	// List	   *children;
	/* TODO: is there any better solution to store children in shared memory? */
	// Oid			children[MAX_PARTITIONS];
	DsmArray    children;
	int			children_count;
	PartType	parttype;
	Index		attnum;
	Oid			atttype;

} PartRelationInfo;

/*
 * Child relation for HASH partitioning
 */
typedef struct HashRelationKey
{
	int		hash;
	Oid		parent_oid;
} HashRelationKey;

typedef struct HashRelation
{
	HashRelationKey key;
	Oid		child_oid;
} HashRelation;

/*
 * Child relation for RANGE partitioning
 */
// typedef struct RangeEntry
// {
// 	Oid			child_oid;
// 	rng_type	min;
// 	rng_type	max;
// } RangeEntry;

typedef struct RangeEntry
{
	Oid		child_oid;
	Datum	min;
	Datum	max;
} RangeEntry;

typedef struct RangeRelation
{
	Oid			parent_oid;
	int			nranges;
	// RangeEntry	ranges[64];
	DsmArray    ranges;
} RangeRelation;


typedef int IndexRange;
#define RANGE_INFINITY 0xFFFF

#define make_range(min, max) \
	((min) << 16 | ((max) & 0x0000FFFF))

#define range_min(range) \
	((range) >> 16)

#define range_max(range) \
	((range) & 0x0000FFFF)

// Range make_range(int min, int max);
// int range_min(Range range);
// int range_max(Range range);
List *append_range(List *a_lst, IndexRange b);
List *intersect_ranges(List *a, List *b);
List *unite_ranges(List *a, List *b);

LWLock *load_config_lock;
LWLock *dsm_init_lock;


// Table *init_dsm_table(size_t block_size);
// DsmArray *alloc_dsm_array(size_t entry_size, size_t length);
// void free_dsm_array(DsmArray *array);
// void *get_dsm_array(const ArrayPtr* ptr);

void alloc_dsm_table();
void create_dsm_segment(size_t block_size);
void init_dsm_table(Table *tbl, dsm_handle h, size_t block_size);
void alloc_dsm_array(DsmArray *arr, size_t entry_size, size_t length);
void free_dsm_array(DsmArray *arr);
void *dsm_array_get_pointer(const DsmArray* arr);


HTAB *relations;
HTAB *hash_restrictions;
HTAB *range_restrictions;
bool initialization_needed;

/* initialization functions */
void init(void);
void create_part_relations_hashtable(void);
void create_hash_restrictions_hashtable(void);
void create_range_restrictions_hashtable(void);
void load_part_relations_hashtable(void);
void load_hash_restrictions(Oid relid);
void load_range_restrictions(Oid relid);
void remove_relation_info(Oid relid);
