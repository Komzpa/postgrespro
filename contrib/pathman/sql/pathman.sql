CREATE EXTENSION pathman;

SET enable_indexscan = OFF;
SET enable_bitmapscan = OFF;

CREATE TABLE hash_rel (
    id      SERIAL PRIMARY KEY,
    value   INTEGER);

SELECT create_hash_partitions('hash_rel', 'value', 3);

INSERT INTO hash_rel VALUES (1, 1);
INSERT INTO hash_rel VALUES (2, 2);
INSERT INTO hash_rel VALUES (3, 3);
INSERT INTO hash_rel VALUES (4, 4);
INSERT INTO hash_rel VALUES (5, 5);
INSERT INTO hash_rel VALUES (6, 6);

EXPLAIN (COSTS OFF) SELECT * FROM hash_rel;
EXPLAIN (COSTS OFF) SELECT * FROM hash_rel WHERE value = 2;
EXPLAIN (COSTS OFF) SELECT * FROM hash_rel WHERE value = 2 OR value = 1;

SELECT drop_hash_partitions('hash_rel');
DROP TABLE hash_rel CASCADE;

CREATE TABLE num_range_rel (
    id SERIAL PRIMARY KEY,
    txt TEXT);
SELECT create_range_partitions('num_range_rel', 'id', 'num', '0', '1000', 3);
INSERT INTO num_range_rel SELECT g, md5(g::TEXT) FROM generate_series(1, 3000) as g;
VACUUM;

EXPLAIN (COSTS OFF) SELECT * FROM num_range_rel WHERE id > 2500;
EXPLAIN (COSTS OFF) SELECT * FROM num_range_rel WHERE id >= 1000 AND id < 3000;
EXPLAIN (COSTS OFF) SELECT * FROM num_range_rel WHERE id >= 1500 AND id < 2500;
EXPLAIN (COSTS OFF) SELECT * FROM num_range_rel WHERE (id >= 500 AND id < 1500) OR (id > 2500);

SELECT drop_range_partitions('num_range_rel');
DROP TABLE num_range_rel CASCADE;

DROP EXTENSION pathman;
