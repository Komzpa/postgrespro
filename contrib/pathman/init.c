#include "pathman.h"
#include "executor/spi.h"
#include "catalog/pg_type.h"
#include "catalog/pg_class.h"
#include "catalog/pg_constraint.h"
#include "catalog/pg_operator.h"
#include "access/htup_details.h"
#include "utils/syscache.h"
#include "utils/builtins.h"
#include "utils/typcache.h"
#include "utils/lsyscache.h"
#include "utils/bytea.h"


HTAB   *relations = NULL;
HTAB   *range_restrictions = NULL;
bool	initialization_needed = true;


static bool validate_range_constraint(Expr *, PartRelationInfo *, Datum *, Datum *);
static bool validate_hash_constraint(Expr *expr, PartRelationInfo *prel, int *hash);
static int cmp_range_entries(const void *p1, const void *p2);
static char *get_extension_schema();

/*
 * Initialize hashtables
 */
void
load_config(void)
{
	bool new_segment_created;

	initialization_needed = false;
	new_segment_created = init_dsm_segment(32);

	LWLockAcquire(load_config_lock, LW_EXCLUSIVE);
	load_relations_hashtable(new_segment_created);
	LWLockRelease(load_config_lock);
}

/*
 * Returns extension schema name or NULL. Caller is responsible for freeing
 * the memory.
 */
static char *
get_extension_schema()
{
	int ret;
	bool isnull;

	ret = SPI_exec("SELECT extnamespace::regnamespace::text FROM pg_extension WHERE extname = 'pathman'", 0);
	if (ret > 0 && SPI_tuptable != NULL)
	{
		TupleDesc tupdesc = SPI_tuptable->tupdesc;
		SPITupleTable *tuptable = SPI_tuptable;
		HeapTuple tuple = tuptable->vals[0];
		Datum datum = SPI_getbinval(tuple, tupdesc, 1, &isnull);

		if (isnull)
			return NULL;

		return TextDatumGetCString(datum);
	}
	return NULL;
}

/*
 * Loads partitioned tables structure to hashtable
 */
void
load_relations_hashtable(bool reinitialize)
{
	int			ret,
				i,
				proc;
	bool		isnull;
	List	   *part_oids = NIL;
	ListCell   *lc;
	char	   *schema;
	PartRelationInfo *prel;
	char		sql[] = "SELECT pg_class.relfilenode, pg_attribute.attnum, pathman_config.parttype, pg_attribute.atttypid "
						"FROM %s.pathman_config "
						"JOIN pg_class ON pg_class.relfilenode = pathman_config.relname::regclass::oid "
						"JOIN pg_attribute ON pg_attribute.attname = pathman_config.attname "
						"AND attrelid = pg_class.relfilenode";
	char *query;

	SPI_connect();
	schema = get_extension_schema();

	/* If extension isn't exist then just quit */
	if (!schema)
	{
		SPI_finish();
		return;
	}

	/* Put schema name to the query */
	query = psprintf(sql, schema);
	ret = SPI_exec(query, 0);
	proc = SPI_processed;

	if (ret > 0 && SPI_tuptable != NULL)
	{
		TupleDesc tupdesc = SPI_tuptable->tupdesc;
		SPITupleTable *tuptable = SPI_tuptable;

		for (i=0; i<proc; i++)
		{
			HeapTuple tuple = tuptable->vals[i];
			int oid = DatumGetObjectId(SPI_getbinval(tuple, tupdesc, 1, &isnull));

			prel = (PartRelationInfo*)
				hash_search(relations, (const void *)&oid, HASH_ENTER, NULL);
			prel->oid = oid;
			prel->attnum = DatumGetInt32(SPI_getbinval(tuple, tupdesc, 2, &isnull));
			prel->parttype = DatumGetInt32(SPI_getbinval(tuple, tupdesc, 3, &isnull));
			prel->atttype = DatumGetObjectId(SPI_getbinval(tuple, tupdesc, 4, &isnull));

			part_oids = lappend_int(part_oids, oid);
		}
	}
	pfree(query);

	/* Load children information */
	foreach(lc, part_oids)
	{
		Oid oid = (int) lfirst_int(lc);

		prel = (PartRelationInfo*)
			hash_search(relations, (const void *)&oid, HASH_FIND, NULL);	

		switch(prel->parttype)
		{
			case PT_RANGE:
				if (reinitialize && prel->children.length > 0)
				{
					RangeRelation *rangerel = (RangeRelation *)
						hash_search(range_restrictions, (void *) &oid, HASH_FIND, NULL);
					free_dsm_array(&prel->children);
					free_dsm_array(&rangerel->ranges);
					prel->children_count = 0;
				}
				load_check_constraints(oid);
				break;
			case PT_HASH:
				if (reinitialize && prel->children.length > 0)
				{
					free_dsm_array(&prel->children);
					prel->children_count = 0;
				}
				load_check_constraints(oid);
				break;
		}
	}
	SPI_finish();
}

void
create_relations_hashtable()
{
	HASHCTL		ctl;

	memset(&ctl, 0, sizeof(ctl));
	ctl.keysize = sizeof(int);
	ctl.entrysize = sizeof(PartRelationInfo);

	/* Already exists, recreate */
	if (relations != NULL)
		hash_destroy(relations);

	relations = ShmemInitHash("Partitioning relation info", 1024, &ctl, HASH_ELEM);
}

/*
 * Load and validate constraints
 */
void
load_check_constraints(Oid parent_oid)
{
	bool		found;
	PartRelationInfo *prel;
	RangeRelation *rangerel;
	int ret;
	int i;
	int proc;

	Datum vals[1];
	Oid oids[1] = {INT4OID};
	bool nulls[1] = {false};
	vals[0] = Int32GetDatum(parent_oid);

	prel = (PartRelationInfo*)
		hash_search(relations, (const void *) &parent_oid, HASH_FIND, &found);

	/* Skip if already loaded */
	if (prel->children.length > 0)
		return;

	ret = SPI_execute_with_args("select pg_constraint.* "
								"from pg_constraint "
								"join pg_inherits on inhrelid = conrelid "
								"where inhparent = $1 and contype='c';",
								1, oids, vals, nulls, true, 0);
	proc = SPI_processed;

	if (ret > 0 && SPI_tuptable != NULL)
	{
		SPITupleTable *tuptable = SPI_tuptable;
		Oid *children;
		RangeEntry *ranges;
		Datum min;
		Datum max;
		int hash;

		alloc_dsm_array(&prel->children, sizeof(Oid), proc);
		children = (Oid *) dsm_array_get_pointer(&prel->children);

		if (prel->parttype == PT_RANGE)
		{
			rangerel = (RangeRelation *)
				hash_search(range_restrictions, (void *) &parent_oid, HASH_ENTER, &found);

			alloc_dsm_array(&rangerel->ranges, sizeof(RangeEntry), proc);
			ranges = (RangeEntry *) dsm_array_get_pointer(&rangerel->ranges);
		}

		for (i=0; i<proc; i++)
		{
			RangeEntry	re;
			HeapTuple	tuple = tuptable->vals[i];
			bool		isnull;
			Datum		val;
			char	   *conbin;
			Expr	   *expr;

			Form_pg_constraint con;

			con = (Form_pg_constraint) GETSTRUCT(tuple);

			val = SysCacheGetAttr(CONSTROID, tuple, Anum_pg_constraint_conbin,
								  &isnull);
			if (isnull)
				elog(ERROR, "null conbin for constraint %u",
					 HeapTupleGetOid(tuple));
			conbin = TextDatumGetCString(val);
			expr = (Expr *) stringToNode(conbin);
			
			switch(prel->parttype)
			{
				case PT_RANGE:
					if (!validate_range_constraint(expr, prel, &min, &max))
					{
						elog(WARNING, "Range constraint for relation %u MUST have exact format: "
									  "VARIABLE >= CONST AND VARIABLE < CONST. Skipping...",
							 (Oid) con->conrelid);
						continue;
					}

					re.child_oid = con->conrelid;
					re.min = min;
					re.max = max;

					ranges[i] = re;
					break;
			
				case PT_HASH:
					if (!validate_hash_constraint(expr, prel, &hash))
					{
						elog(WARNING, "Hash constraint for relation %u MUST have exact format: "
									  "VARIABLE %% CONST = CONST. Skipping...",
							 (Oid) con->conrelid);
						continue;
					}
					children[hash] = con->conrelid;
			}
		}
		prel->children_count = proc;

		if (prel->parttype == PT_RANGE)
		{
			/* Sort ascending */
			qsort(ranges, proc, sizeof(RangeEntry), cmp_range_entries);

			/* Copy oids to prel */
			for(i=0; i < proc; i++)
				children[i] = ranges[i].child_oid;
		}

		/* Check if some ranges overlap */
		for(i=0; i < proc-1; i++)
		{
			if (ranges[i].max > ranges[i+1].min)
			{
				elog(WARNING, "Partitions %u and %u overlap. Disabling pathman for relation %u..",
					 ranges[i].child_oid, ranges[i+1].child_oid, parent_oid);
				hash_search(relations, (const void *) &parent_oid, HASH_REMOVE, &found);
			}
		}
	}
}


/* qsort comparison function for oids */
static int
cmp_range_entries(const void *p1, const void *p2)
{
	const RangeEntry	*v1 = (const RangeEntry *) p1;
	const RangeEntry	*v2 = (const RangeEntry *) p2;

	if (v1->min < v2->min)
		return -1;
	if (v1->min > v2->min)
		return 1;
	return 0;
}

/*
 * Validates range constraint. It MUST have the exact format:
 * VARIABLE >= CONST AND VARIABLE < CONST
 */
static bool
validate_range_constraint(Expr *expr, PartRelationInfo *prel, Datum *min, Datum *max)
{
	TypeCacheEntry *tce;
	BoolExpr *boolexpr = (BoolExpr *) expr;
	OpExpr *opexpr;

	/* it should be an AND operator on top */
	if ( !(IsA(expr, BoolExpr) && boolexpr->boolop == AND_EXPR) )
		return false;

	/* and it should have exactly two operands */
	if (list_length(boolexpr->args) != 2)
		return false;

	tce = lookup_type_cache(prel->atttype, TYPECACHE_EQ_OPR | TYPECACHE_LT_OPR | TYPECACHE_GT_OPR);

	/* check that left operand is >= operator */
	opexpr = (OpExpr *) linitial(boolexpr->args);
	if (get_op_opfamily_strategy(opexpr->opno, tce->btree_opf) == BTGreaterEqualStrategyNumber)
	{
		Node *left = linitial(opexpr->args);
		Node *right = lsecond(opexpr->args);
		if ( !IsA(left, Var) || !IsA(right, Const) )
			return false;
		if ( ((Var*) left)->varattno != prel->attnum )
			return false;
		*min = ((Const*) right)->constvalue;
	}
	else
		return false;

	/* TODO: rewrite this */
	/* check that right operand is < operator */
	opexpr = (OpExpr *) lsecond(boolexpr->args);
	if (get_op_opfamily_strategy(opexpr->opno, tce->btree_opf) == BTLessStrategyNumber)
	{
		Node *left = linitial(opexpr->args);
		Node *right = lsecond(opexpr->args);
		if ( !IsA(left, Var) || !IsA(right, Const) )
			return false;
		if ( ((Var*) left)->varattno != prel->attnum )
			return false;
		*max = ((Const*) right)->constvalue;
	}
	else
		return false;

	return true;
}

/*
 * Validate hash constraint. It MUST have the exact format
 * VARIABLE % CONST = CONST
 */
static bool
validate_hash_constraint(Expr *expr, PartRelationInfo *prel, int *hash)
{
	OpExpr *eqexpr;
	OpExpr *modexpr;

	if (!IsA(expr, OpExpr))
		return false;
	eqexpr = (OpExpr *) expr;

	/* Is this an equality operator? */
	if (eqexpr->opno != Int4EqualOperator)
		return false;

	if (!IsA(linitial(eqexpr->args), OpExpr))
		return false;

	/* Is this a modulus operator? */
	modexpr = (OpExpr *) linitial(eqexpr->args);
	if (modexpr->opno != 530)
		return false;

	if (list_length(modexpr->args) == 2)
	{
		Node *left = linitial(modexpr->args);
		Node *right = lsecond(modexpr->args);
		Const *mod_result;

		if ( !IsA(left, Var) || !IsA(right, Const) )
			return false;
		if ( ((Var*) left)->varattno != prel->attnum )
			return false;
		if (DatumGetInt32(((Const*) right)->constvalue) != prel->children.length)
			return false;

		if ( !IsA(lsecond(eqexpr->args), Const) )
			return false;

		mod_result = lsecond(eqexpr->args);
		*hash = DatumGetInt32(mod_result->constvalue);
		return true;
	}

	return false;
}

/*
 * Create range restrictions table
 */
void
create_range_restrictions_hashtable()
{
	HASHCTL		ctl;

	memset(&ctl, 0, sizeof(ctl));
	ctl.keysize = sizeof(int);
	ctl.entrysize = sizeof(RangeRelation);
	range_restrictions = ShmemInitHash("pg_pathman range restrictions",
									   1024, &ctl, HASH_ELEM | HASH_BLOBS);
}

/*
 * Remove partitions
 */
void
remove_relation_info(Oid relid)
{
	PartRelationInfo   *prel;
	RangeRelation	   *rangerel;

	prel = (PartRelationInfo *)
		hash_search(relations, (const void *) &relid, HASH_FIND, 0);

	/* If there is nothing to remove then just return */
	if (!prel)
		return;

	/* Remove children relations */
	switch (prel->parttype)
	{
		case PT_HASH:
			free_dsm_array(&prel->children);
			break;
		case PT_RANGE:
			rangerel = (RangeRelation *)
				hash_search(range_restrictions, (const void *) &relid, HASH_FIND, 0);
			free_dsm_array(&rangerel->ranges);
			free_dsm_array(&prel->children);
			hash_search(range_restrictions, (const void *) &relid, HASH_REMOVE, 0);
			break;
	}
	prel->children_count = 0;
	hash_search(relations, (const void *) &relid, HASH_REMOVE, 0);
}