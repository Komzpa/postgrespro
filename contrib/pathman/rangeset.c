#include "pathman.h"

bool
irange_intersects(IndexRange a, IndexRange b)
{
	return (irange_lower(a) <= irange_upper(b)) ||
		   (irange_lower(b) <= irange_upper(a));
}

bool
irange_conjuncted(IndexRange a, IndexRange b)
{
	return (irange_lower(a) - 1 <= irange_upper(b)) ||
		   (irange_lower(b) - 1 <= irange_upper(a));
}

IndexRange
irange_union(IndexRange a, IndexRange b)
{
	Assert(irange_is_lossy(a) == irange_is_lossy(b));
	return make_irange(Min(irange_lower(a), irange_lower(b)),
					   Max(irange_upper(a), irange_upper(b)),
					   irange_is_lossy(a));
}

IndexRange
irange_intersect(IndexRange a, IndexRange b)
{
	return make_irange(Max(irange_lower(a), irange_lower(b)),
					   Min(irange_upper(a), irange_upper(b)),
					   irange_is_lossy(a) || irange_is_lossy(b));
}

List *
irange_list_union(List *a, List *b)
{
	ListCell   *ca,
			   *cb;
	List	   *result = NIL;
	IndexRange	cur;
	bool		have_cur = false;

	ca = list_head(a);
	cb = list_head(b);

	while (ca || cb)
	{
		IndexRange	next;

		if (ca && cb)
		{
			if (irange_lower(lfirst_irange(ca)) <= irange_lower(lfirst_irange(cb)))
				next = lfirst_irange(ca);
			else
				next = lfirst_irange(cb);
		}
		else if (ca)
		{
			next = lfirst_irange(ca);
		}
		else if (cb)
		{
			next = lfirst_irange(cb);
		}

		if (!have_cur)
		{
			cur = next;
			have_cur = true;
		}
		else
		{
			if (irange_conjuncted(next, cur))
			{
				if (irange_is_lossy(next) == irange_is_lossy(cur))
				{
					cur = irange_union(next, cur);
				}
				else
				{
					if (!irange_is_lossy(cur))
					{
						result = lappend_irange(result, cur);
						cur = make_irange(irange_upper(cur) + 1,
												 irange_upper(next),
												 irange_is_lossy(next));
					}
					else
					{
						result = lappend_irange(result, 
									make_irange(irange_lower(cur),
												irange_lower(next) - 1,
												irange_is_lossy(cur)));
						cur = next;
					}
				}
			}
			else
			{
				result = lappend_irange(result, cur);
				cur = next;
			}
		}
	}

	if (have_cur)
		result = lappend_irange(result, cur);

	return result;
}

List *
irange_list_intersect(List *a, List *b)
{
	ListCell   *ca,
			   *cb;
	List	   *result = NIL;
	IndexRange	ra, rb;

	ca = list_head(a);
	cb = list_head(b);

	while (ca && cb)
	{
		ra = lfirst_irange(ca);
		rb = lfirst_irange(cb);
		if (irange_intersects(ra, rb))
			result = lappend_irange(result, irange_intersect(ra, rb));

		if (irange_lower(ra) <= irange_lower(rb))
			ca = lnext(ca);
		else
			cb = lnext(cb);
	}
	return result;
}
