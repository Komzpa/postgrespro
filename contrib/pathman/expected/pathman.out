CREATE EXTENSION pathman;
SET enable_indexscan = OFF;
SET enable_bitmapscan = OFF;
CREATE TABLE hash_rel (
    id      SERIAL PRIMARY KEY,
    value   INTEGER);
SELECT create_hash_partitions('hash_rel', 'value', 3);
NOTICE:  trigger "hash_rel_insert_trigger" for relation "hash_rel" does not exist, skipping
NOTICE:  function hash_rel_hash_insert_trigger_func() does not exist, skipping
NOTICE:  trigger "hash_rel_update_trigger" for relation "hash_rel" does not exist, skipping
NOTICE:  function hash_rel_hash_update_trigger_func() does not exist, skipping
 create_hash_partitions 
------------------------
 
(1 row)

INSERT INTO hash_rel VALUES (1, 1);
INSERT INTO hash_rel VALUES (2, 2);
INSERT INTO hash_rel VALUES (3, 3);
INSERT INTO hash_rel VALUES (4, 4);
INSERT INTO hash_rel VALUES (5, 5);
INSERT INTO hash_rel VALUES (6, 6);
EXPLAIN (COSTS OFF) SELECT * FROM hash_rel;
          QUERY PLAN          
------------------------------
 Append
   ->  Seq Scan on hash_rel_0
   ->  Seq Scan on hash_rel_1
   ->  Seq Scan on hash_rel_2
(4 rows)

EXPLAIN (COSTS OFF) SELECT * FROM hash_rel WHERE value = 2;
          QUERY PLAN          
------------------------------
 Append
   ->  Seq Scan on hash_rel_2
         Filter: (value = 2)
(3 rows)

EXPLAIN (COSTS OFF) SELECT * FROM hash_rel WHERE value = 2 OR value = 1;
          QUERY PLAN          
------------------------------
 Append
   ->  Seq Scan on hash_rel_1
         Filter: (value = 1)
   ->  Seq Scan on hash_rel_2
         Filter: (value = 2)
(5 rows)

SELECT drop_hash_partitions('hash_rel');
NOTICE:  trigger "hash_rel_update_trigger" for relation "hash_rel" does not exist, skipping
NOTICE:  function hash_rel_hash_update_trigger_func() does not exist, skipping
 drop_hash_partitions 
----------------------
 
(1 row)

DROP TABLE hash_rel CASCADE;
NOTICE:  drop cascades to 3 other objects
DETAIL:  drop cascades to table hash_rel_0
drop cascades to table hash_rel_1
drop cascades to table hash_rel_2
CREATE TABLE num_range_rel (
    id SERIAL PRIMARY KEY,
    txt TEXT);
SELECT create_range_partitions('num_range_rel', 'id', 'num', '0', '1000', 3);
 create_range_partitions 
-------------------------
 
(1 row)

INSERT INTO num_range_rel SELECT g, md5(g::TEXT) FROM generate_series(1, 3000) as g;
VACUUM;
EXPLAIN (COSTS OFF) SELECT * FROM num_range_rel WHERE id > 2500;
              QUERY PLAN              
--------------------------------------
 Append
   ->  Seq Scan on num_range_rel_2000
         Filter: (id > 2500)
   ->  Seq Scan on num_range_rel_3000
(4 rows)

EXPLAIN (COSTS OFF) SELECT * FROM num_range_rel WHERE id >= 1000 AND id < 3000;
              QUERY PLAN              
--------------------------------------
 Append
   ->  Seq Scan on num_range_rel_1000
   ->  Seq Scan on num_range_rel_2000
(3 rows)

EXPLAIN (COSTS OFF) SELECT * FROM num_range_rel WHERE id >= 1500 AND id < 2500;
              QUERY PLAN              
--------------------------------------
 Append
   ->  Seq Scan on num_range_rel_1000
         Filter: (id >= 1500)
   ->  Seq Scan on num_range_rel_2000
         Filter: (id < 2500)
(5 rows)

EXPLAIN (COSTS OFF) SELECT * FROM num_range_rel WHERE (id >= 500 AND id < 1500) OR (id > 2500);
              QUERY PLAN              
--------------------------------------
 Append
   ->  Seq Scan on num_range_rel_0
         Filter: (id >= 500)
   ->  Seq Scan on num_range_rel_1000
         Filter: (id < 1500)
   ->  Seq Scan on num_range_rel_2000
         Filter: (id > 2500)
   ->  Seq Scan on num_range_rel_3000
(8 rows)

SELECT drop_range_partitions('num_range_rel');
 drop_range_partitions 
-----------------------
 
(1 row)

DROP TABLE num_range_rel CASCADE;
DROP EXTENSION pathman;
